feat(rbac): add support for creation of role

divyanshiGupta opened the pull request:
Story: https://github.com/janus-idp/backstage-plugins/issues/941, https://github.com/janus-idp/backstage-plugins/issues/942

This PR adds support for creating a role via a create role form. It adds a create button on roles list page which takes the user to the create role form. User can provide name and description and also add  members (users/groups) to the role. On successful creation user is redirected to the roles list page.

Remaining ACs will be covered in this story - https://github.com/janus-idp/backstage-plugins/issues/975


https://github.com/janus-idp/backstage-plugins/assets/20724543/32437302-e95f-4f72-9554-4c5325587a91



Error on submit:
<img width="1792" alt="Screenshot 2023-11-30 at 9 15 21 PM" src="https://github.com/janus-idp/backstage-plugins/assets/20724543/14d54088-cbf9-4a9e-a60a-ebe457df73f3">



To Do

- [x] Add validation
- [x] Add tests
- [x] Clean up css

debsmita1 commented:
The create role review is missing

![Screenshot 2023-11-29 at 11 46 35 PM](https://github.com/janus-idp/backstage-plugins/assets/22490998/d5283e38-4fe0-4ea8-ba02-d78b72b95f77)



The delete icon is missing from the table

![Screenshot 2023-11-29 at 11 46 17 PM](https://github.com/janus-idp/backstage-plugins/assets/22490998/eb1076c9-e18b-451f-aefb-b751cdf2a802)



The empty state needs to be fixed

![Screenshot 2023-11-29 at 11 46 08 PM](https://github.com/janus-idp/backstage-plugins/assets/22490998/b1e8932b-d2fb-48f6-9ae5-c6cbe6dc116c)


On hitting create and on successful creation user should be taken to the roles list page

debsmita1 commented:
Would you be adding the permission check for creating roles as part of this issue https://github.com/janus-idp/backstage-plugins/issues/975 ?

debsmita1 commented:
The Back and Reset buttons in the Figma are outlined,

divyanshiGupta commented:
> The Back and Reset buttons in the Figma are outlined,

@debsmita1 the figma is based on new theme but for now I have kept everything consistent with what we already have. @parvathyvr lmk wdyt.

parvathyvr commented:
> > The Back and Reset buttons in the Figma are outlined,
> 
> @debsmita1 the figma is based on new theme but for now I have kept everything consistent with what we already have. @parvathyvr lmk wdyt.

If we are using the existing theme, i am okay with having the button style in the current implementation.
1 change- Step 2 should be 'Add users ad groups'.

gashcrumb commented on plugins/rbac/dev/index.tsx:
```diff
@@ -76,6 +76,8 @@ class MockRBACApi implements RBACAPI {
     return 204;
   }
+
+  async createRole(): Promise<any> {}
```
Is there a better type than `any` for this?

divyanshiGupta commented on plugins/rbac/src/components/CreateRole/CreateRoleForm.test.tsx:
```diff
@@ -0,0 +1,42 @@
+  it('handles form submission failure', async () => {
+    render(<CreateRoleForm />);
+
+    expect(screen.getByText(/unable to create role/i)).toBeInTheDocument();
```
@gashcrumb On submission if there is any error, formik status gets updated with that error.
```
{formik.status?.submitError && (
      <Box>
          <Alert severity="error">{`Unable to create role. ${formik.status.submitError}`}</Alert>
       </Box>
  )}
```
So here I was just trying to test if any such error is present in the status, the error message is displayed or not. I have mocked formik and returning the status with an error to test the above condition.  I will update the test so that its not confusing.

divyanshiGupta commented:
@parvathyvr @debsmita1 for now I have added the following permission checks on accessibility of `Create` role button:

- policy.entity.create
- catalog.entity.read
- catalog.entity.create

If the user has all the above permissions then they can go ahead and create the role, otherwise the button is disabled. LMK if  anything else is required here.

divyanshiGupta commented:
> > @parvathyvr @debsmita1 for now I have added the following permission checks on accessibility of `Create` role button:
> > 
> > * policy.entity.create
> > * catalog.entity.read
> > * catalog.entity.create
> > 
> > If the user has all the above permissions then they can go ahead and create the role, otherwise the button is disabled. LMK if anything else is required here.
> 
> @divyanshiGupta `catalog.entity.create` this won't be necessary, we only want to fetch the users and groups

@debsmita1 I have added `catalog.entity.create` because during role creation we make an api call for creating the role. LMK if we dont need this check for that or `policy.entity.create` is sufficient?

divyanshiGupta commented:
> > > @parvathyvr @debsmita1 for now I have added the following permission checks on accessibility of `Create` role button:
> > > 
> > > * policy.entity.create
> > > * catalog.entity.read
> > > * catalog.entity.create
> > > 
> > > If the user has all the above permissions then they can go ahead and create the role and policies to it, otherwise the button is disabled. LMK if anything else is required here.
> > 
> > 
> > @divyanshiGupta `catalog.entity.create` this won't be necessary, we only want to fetch the users and groups
> 
> @debsmita1 I have added `catalog.entity.create` because during role creation we make an api call for creating the role as well. LMK if we dont need this check for that or `policy.entity.create` is sufficient?

@PatAKnight wdyt?

divyanshiGupta commented:
> > > The Back and Reset buttons in the Figma are outlined,
> > 
> > 
> > @debsmita1 the figma is based on new theme but for now I have kept everything consistent with what we already have. @parvathyvr lmk wdyt.
> 
> If we are using the existing theme, i am okay with having the button style in the current implementation. Few things to update -Step 2 should be 'Add users and groups'. -Also i notice we are using 'Create a Role' and 'Create new role'. I think let's go with 'Create role' in the header and in the stepper so that it will be consistent when we have 'Edit role'.

@parvathyvr Done.
<img width="1792" alt="Screenshot 2023-12-06 at 7 28 20 PM" src="https://github.com/janus-idp/backstage-plugins/assets/20724543/99825547-ba5c-44e3-b22d-de349de3a57c">

debsmita1 commented on plugins/rbac-common/src/permissions.ts:
```diff
@@ -67,3 +69,34 @@ export const policyEntityPermissions = [
+ * This permission is used to authorize actions that involve reading
+ * catalog entities.
+ */
+export const catalogEntityReadPermission = createPermission({
```
you do not need to create this

you can import from the catalog-common plugin

we are doing this in App.tsx
```
import { catalogEntityReadPermission } from '@backstage/plugin-catalog-common/alpha';
```

debsmita1 commented on plugins/rbac-common/src/permissions.ts:
```diff
@@ -67,3 +69,34 @@ export const policyEntityPermissions = [
+/**
+ * This permission is used to authorize the creation of new catalog entities.
+ */
+export const catalogEntityCreatePermission = createPermission({
```
Let's remove this as we do not need create

debsmita1 commented:
We should disable the route if a user is not authorized to create roles


https://github.com/janus-idp/backstage-plugins/assets/22490998/e263900e-faec-4e54-9159-53350dd0e5d6

debsmita1 commented on plugins/rbac/src/components/CreateRole/CreateRolePage.tsx:
```diff
@@ -0,0 +1,16 @@
+
+export const CreateRolePage = () => {
+  return (
+    <Page themeId="tool">
+      <Header title="Create role" />
+      <Content>
+        <CreateRoleForm />
+      </Content>
+    </Page>
```
```suggestion
<RequirePermission
      permission={policyEntityCreatePermission}
      resourceRef={policyEntityCreatePermission.resourceType}
    >
    <Page themeId="tool">
      <Header title="Create role" />
      <Content>
        <CreateRoleForm />
      </Content>
    </Page>
    </RequirePermission>
```

debsmita1 commented on plugins/rbac/src/components/CreateRole/AddMembersForm.tsx:
```diff
@@ -0,0 +1,136 @@
+              </span>
+            </Box>
+          );
+        }}
```
Can you move this to a render component?

debsmita1 commented:
I can add a user or group multiple times, and in this scenario, the backend does not generate an error
cc @PatAKnight @AndrienkoAleksandr

divyanshiGupta commented:
> I can add a user or group multiple times, and in this scenario, the backend does not generate an error cc @PatAKnight @AndrienkoAleksandr

If there is a role with some users/groups and then an api call is made to add same user/group to the role then the backend returns an error but if we add same user/group multiple times in a single api call then the role successfully gets created. To avoid this from UI I have disabled the members that are already selected and can be seen in the table. 

https://github.com/janus-idp/backstage-plugins/assets/20724543/00a8e50e-7673-4a91-824f-74ddb04f1689

cc: @parvathyvr @debsmita1

debsmita1 commented on plugins/rbac/src/components/CreateRole/CreateRolePage.tsx:
```diff
@@ -0,0 +1,23 @@
+  return (
+    <RequirePermission
+      permission={catalogEntityReadPermission}
+      resourceRef={catalogEntityReadPermission.resourceType}
```
Should check for catalogEntityCreatePermission

divyanshiGupta replied:
> > > > @parvathyvr @debsmita1 for now I have added the following permission checks on accessibility of `Create` role button:
> > > > 
> > > > * policy.entity.create
> > > > * catalog.entity.read
> > > > * catalog.entity.create
> > > > 
> > > > If the user has all the above permissions then they can go ahead and create the role and policies to it, otherwise the button is disabled. LMK if anything else is required here.
> > > 
> > > 
> > > @divyanshiGupta `catalog.entity.create` this won't be necessary, we only want to fetch the users and groups
> > 
> > 
> > @debsmita1 I have added `catalog.entity.create` because during role creation we make an api call for creating the role as well. LMK if we dont need this check for that or `policy.entity.create` is sufficient?
> 
> for creating the role we need policy.entity.create

@debsmita1 I am confused. Do we need catalog.entity.create for creating roles or not?

debsmita1 replied:
Sorry, my bad
I was looking for policyEntityCreate & catalogEntityRead .Found the 2 places where you have added the permission checks

gashcrumb commented on plugins/rbac/src/components/RolesList.tsx:
```diff
@@ -61,6 +66,17 @@ export const RolesList = () => {
           {toastMessage}
         </Alert>
       </Snackbar>
+      <span className={classes.createRoleButton}>
```
Frustrating that the backstage components seem to avoid having a `className` property.  Anyway, it's really only a nit but I suggest maybe call this class `createRoleButtonWrapper` just so it's more clear earlier in the file where this classname is defined that it doesn't target the button but instead a wrapper component that's used to align the button.

gashcrumb commented on plugins/rbac/src/components/CreateRole/RoleDetailsForm.tsx:
```diff
@@ -0,0 +1,50 @@
+import { CreateRoleFormValues } from './types';
+
+export const RoleDetailsForm = ({
+  formik,
```
Here is another example where I would suggest not accepting the formik instance but instead accept the specific properties.   Let the parent component deal with the formik instance, this component then becomes much simpler and potentially usable with other form libraries even.

gashcrumb commented on plugins/rbac/src/components/CreateRole/AddMembersForm.tsx:
```diff
@@ -0,0 +1,116 @@
+
+import { useApi } from '@backstage/core-plugin-api';
+
+import { LinearProgress, TextField } from '@material-ui/core';
```
I mentioned this on @debsmita1's PR but probably good to mention it here too.  I think we should try and use Material 5 whenever possible, so `@mui/material/LinearProgress` , `@mui/material/TextField` and so on.  It could be too late to do this for this particular iteration and that's fine, we'll have to do it eventually when upstream drops Material 4.

invincibleJai commented:
@divyanshiGupta dev mode doesn't seem to work

<img width="1728" alt="image" src="https://github.com/janus-idp/backstage-plugins/assets/5129024/598eb1c8-16ea-43cf-a6d6-a46c8b324ca2">

invincibleJai commented:
@divyanshiGupta Thanks for the work, overall looks good. I have one observation

Create btn is disabled, and no info is shown for the reason though the user has create access to the policy entity 

<img width="1728" alt="image" src="https://github.com/janus-idp/backstage-plugins/assets/5129024/dc69cdb4-b409-4d99-a94c-7c64d0ff644d">
 
I believe that is because the logged-in user didn't have create permission to `catalog-entity` so README should cover that or a tooltip

invincibleJai commented on plugins/rbac/src/components/CreateRole/RoleDetailsForm.tsx:
```diff
@@ -0,0 +1,59 @@
+      <TextField
+        label="Description"
+        variant="outlined"
+        helperText="Enter a brief description about the role(The purpose of the role)"
```
Nit: this text is per UXD?

```suggestion
        helperText="Enter a brief description about the role (The purpose of the role)"
```

divyanshiGupta replied:
@invincibleJai Yes it is as per https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=932-1542&mode=design&t=V40DvGtSyENzJFNB-4.

invincibleJai replied:
so there is space :)

divyanshiGupta replied:
yes, added the space.

invincibleJai reviewed (approved):
LGTM

divyanshiGupta commented:
> @divyanshiGupta Thanks for the work, overall looks good. I have one observation
> 
> Create btn is disabled, and no info is shown for the reason though the user has create access to the policy entity
> 
> <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/5129024/289544687-dc69cdb4-b409-4d99-a94c-7c64d0ff644d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDI1MTEsIm5iZiI6MTcwMjMwMjIxMSwicGF0aCI6Ii81MTI5MDI0LzI4OTU0NDY4Ny1kYzY5Y2RiNC1iNDA5LTRkOTktYTk0Yy03YzY0ZDBmZjY0NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMTM0MzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTI2NDcwZTE1NWUwNmZlODI5NjRmNzE1ZWFmZGYwYTU5NjJhMjQxNGQyMmJiZjM0ZDVmNDhkNmRlZmU1MzNiYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.INDLwKHRDz0Uv_U_Qva6ytB-CRg5OO_Lrk-7peJozJQ">
> I believe that is because the logged-in user didn't have create permission to `catalog-entity` so README should cover that or a tooltip

@invincibleJai yes it need `policy.entity.create` and `catalog.entity.read` permissions. Good idea to add in README.

gashcrumb reviewed (approved):
:+1:

parvathyvr commented:
> @divyanshiGupta Thanks for the work, overall looks good. I have one observation
> 
> Create btn is disabled, and no info is shown for the reason though the user has create access to the policy entity
> <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/5129024/289544687-dc69cdb4-b409-4d99-a94c-7c64d0ff644d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDM3NjksIm5iZiI6MTcwMjMwMzQ2OSwicGF0aCI6Ii81MTI5MDI0LzI4OTU0NDY4Ny1kYzY5Y2RiNC1iNDA5LTRkOTktYTk0Yy03YzY0ZDBmZjY0NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMTQwNDI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzliZThmMjRiNTgwOTcyYjBlZTgxNzcyYmQ5N2Y5NGIzNDM2OWQ0NTZhMGM4OGRhZDY2YzNmYjA2YTJhYzMwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AYu50VraI-UqmkR5xM2akv-6FvMaTJDAX7wC8mg3RGw">
> 
> I believe that is because the logged-in user didn't have create permission to `catalog-entity` so README should cover that or a tooltip

Hi @divyanshiGupta I have added an inline alert scenario if 'Create' button is disabled- https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=1684-572&mode=design&t=4ygFyKpEzJg5P0h6-4
(The alert message is for a different scenario- but i assume we can use this pattern to display to the user as to why 'Create' is disabled in the UI).

invincibleJai commented:
> > @divyanshiGupta Thanks for the work, overall looks good. I have one observation
> > Create btn is disabled, and no info is shown for the reason though the user has create access to the policy entity
> > <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/5129024/289544687-dc69cdb4-b409-4d99-a94c-7c64d0ff644d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDM3NjksIm5iZiI6MTcwMjMwMzQ2OSwicGF0aCI6Ii81MTI5MDI0LzI4OTU0NDY4Ny1kYzY5Y2RiNC1iNDA5LTRkOTktYTk0Yy03YzY0ZDBmZjY0NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMTQwNDI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzliZThmMjRiNTgwOTcyYjBlZTgxNzcyYmQ5N2Y5NGIzNDM2OWQ0NTZhMGM4OGRhZDY2YzNmYjA2YTJhYzMwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AYu50VraI-UqmkR5xM2akv-6FvMaTJDAX7wC8mg3RGw">
> > I believe that is because the logged-in user didn't have create permission to `catalog-entity` so README should cover that or a tooltip
> 
> Hi @divyanshiGupta I have added an inline alert scenario if 'Create' button is disabled- https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=1684-572&mode=design&t=4ygFyKpEzJg5P0h6-4 (The alert message is for a different scenario- but i assume we can use this pattern to display to the user as to why 'Create' is disabled in the UI).

@parvathyvr thanks 👍 

@divyanshiGupta would you like to take this as follow up i.e as part of your final create flow PR?

divyanshiGupta commented:
> > > @divyanshiGupta Thanks for the work, overall looks good. I have one observation
> > > Create btn is disabled, and no info is shown for the reason though the user has create access to the policy entity
> > > <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/5129024/289544687-dc69cdb4-b409-4d99-a94c-7c64d0ff644d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIzMDM3NjksIm5iZiI6MTcwMjMwMzQ2OSwicGF0aCI6Ii81MTI5MDI0LzI4OTU0NDY4Ny1kYzY5Y2RiNC1iNDA5LTRkOTktYTk0Yy03YzY0ZDBmZjY0NGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTFUMTQwNDI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzliZThmMjRiNTgwOTcyYjBlZTgxNzcyYmQ5N2Y5NGIzNDM2OWQ0NTZhMGM4OGRhZDY2YzNmYjA2YTJhYzMwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AYu50VraI-UqmkR5xM2akv-6FvMaTJDAX7wC8mg3RGw">
> > > I believe that is because the logged-in user didn't have create permission to `catalog-entity` so README should cover that or a tooltip
> > 
> > 
> > Hi @divyanshiGupta I have added an inline alert scenario if 'Create' button is disabled- https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=1684-572&mode=design&t=4ygFyKpEzJg5P0h6-4 (The alert message is for a different scenario- but i assume we can use this pattern to display to the user as to why 'Create' is disabled in the UI).
> 
> @parvathyvr thanks 👍
> 
> @divyanshiGupta would you like to take this as follow up i.e as part of your final create flow PR?

@invincibleJai sure  I will address this in my next PR along with updating the README (based on what we decide if the change should be in rbac backend plugin or UI).

debsmita1 commented:
LGTM
